lib/BuildSystem/LaneBasedExecutionQueue.cpp

```diff
@@ -281,26 +295,88 @@ class LaneBasedExecutionQueue : public BuildExecutionQueue {
+        getDelegate().commandProcessHadError(
+          context.job.getForCommand(), handle,
+          Twine("unable to set the non-blocking flag in stdout pipe (") + strerror(errno) + ")");
+        getDelegate().commandProcessFinished(context.job.getForCommand(), handle,
+                                           -1);
+        return false;
+      }
+      flags = fcntl(stderrPipe[0], F_GETFL, 0);
+      if (fcntl(stderrPipe[0], F_SETFL, flags | O_NONBLOCK)) {
+        getDelegate().commandProcessHadError(
+          context.job.getForCommand(), handle,
+          Twine("unable to set the non-blocking flag in stderr pipe (") + strerror(errno) + ")");
+        getDelegate().commandProcessFinished(context.job.getForCommand(), handle,
+                                           -1);
+        return false;
+      }
+     
+      // We use select() to determine which fd has output ready to be read. Once select()
+      // returns we read whichever (or both) fd has data available in non-blocking mode.
+      // We stop watching a fd once it has read all of the data i.e. read() returns 0.
+      // When all of the fds are done being watched we can exit the loop.
+      bool watchingStdout = true, watchingStderr = true;
+      while(true) {
+        fd_set readSet;
+        FD_ZERO(&readSet);
+        int maxFD = 0;
+        // If we're watching a fd, add it to the readSet and determine the maxFD for select().
+        if (watchingStdout) { 
+          FD_SET(outputPipe[0], &readSet); 
+          maxFD = outputPipe[0]; 
+        }
+        if (watchingStderr) { 
+          FD_SET(stderrPipe[0], &readSet); 
+          maxFD = maxFD > stderrPipe[0] ? maxFD : stderrPipe[0]; 
+        }
+        
+        int selectResult = select(maxFD + 1, &readSet, nullptr, nullptr, nullptr);
+        assert(selectResult != 0); 
+
+        if (selectResult < 0) {
+          if (errno == EINTR) continue; // Interrupted. Try again.
           getDelegate().commandProcessHadError(
               context.job.getForCommand(), handle,
-              Twine("unable to read process output (") + strerror(errno) + ")");
+              Twine("unable to select() while reading from subprocess (") + strerror(errno) + ")");
           break;
         }
 
-        if (numBytes == 0)
-          break;
+        // Read the FD and stop watching it if all the data has been read.
+        // If after a read we're still watching that fd, make sure we got EWOULDBLOCK from read().
+        if (FD_ISSET(outputPipe[0], &readSet)) {
+          FD_CLR(outputPipe[0], &readSet);
+          watchingStdout = !readFd(outputPipe[0], hasPartialOutput);
+          if (watchingStdout && errno != EWOULDBLOCK) { 
+            getDelegate().commandProcessHadError(
+                context.job.getForCommand(), handle,
+                Twine("unable to select() while reading from subprocess (") + strerror(errno) + ")");
+            break;
+          }
```

Comment:
I think I would just make readFd report the errors, since it is the thing which is pulling from the `fd` and knows what errors are possible.